File: cyrio-java/src/main/java/com/cyrio/core/model/StorageInfo.java
package com.cyrio.core.model;

/**
 * 存储空间信息（用户面向数据模型）
 *
 * <p>由 Rust cyrio-core 通过 JNI 返回的 JSON 数据解析而来。
 * 描述一个内存单元（内置闪存或 SD 卡）的容量信息。
 * S-Series 中所有大小字段单位是字节，显示时 /1024/1024 得 MB。
 *
 * <p>若请求不存在的内存单元（如未插 SD 卡时查询单元 1），
 * Rust 端返回全 0，{@code isPresent} 为 {@code false}。
 */
public class StorageInfo {

    /** 总大小（字节） */
    public long totalSize;

    /** 已用字节数 */
    public long usedSize;

    /** 空闲字节数 */
    public long freeSize;

    /** 系统保留字节数 */
    public long systemSize;

    /** 内存单元名（如 "Internal Memory"） */
    public String name;

    /** 型号字符串 */
    public String model;

    /** 该内存单元是否存在（size > 0） */
    public boolean isPresent;

    /**
     * 创建一个空的 {@code StorageInfo}
     */
    public StorageInfo() {
        this.totalSize = 0;
        this.usedSize = 0;
        this.freeSize = 0;
        this.systemSize = 0;
        this.name = "";
        this.model = "";
        this.isPresent = false;
    }

    /**
     * 返回内存单元大小的 MB 字符串（人类可读）
     *
     * @return 如 "10.0MB used / 50.0MB free / 64.0MB total"
     */
    public String formatSize() {
        double totalMb = totalSize / (1024.0 * 1024.0);
        double freeMb = freeSize / (1024.0 * 1024.0);
        double usedMb = usedSize / (1024.0 * 1024.0);
        return String.format("%.1fMB used / %.1fMB free / %.1fMB total", usedMb, freeMb, totalMb);
    }

    @Override
    public String toString() {
        return "StorageInfo{"
                + "totalSize=" + totalSize
                + ", usedSize=" + usedSize
                + ", freeSize=" + freeSize
                + ", systemSize=" + systemSize
                + ", name='" + name + "'"
                + ", model='" + model + "'"
                + ", isPresent=" + isPresent
                + "}";
    }
}

File: platforms/android-native/app/src/main/java/c/cyrio/android/model/StorageInfo.kt
package c.cyrio.android.model

import org.json.JSONObject

/**
 * 存储信息模型
 *
 * 对应 Rust 侧 StorageJson：
 * {"totalSize","usedSize","freeSize","systemSize","name","model","isPresent"}
 */
data class StorageInfo(
    val totalSize: Int,
    val usedSize: Int,
    val freeSize: Int,
    val systemSize: Int,
    val name: String,
    val model: String,
    val isPresent: Boolean
) {
    /** 已用百分比 */
    val usedPercent: Int
        get() = if (totalSize > 0) (usedSize * 100 / totalSize) else 0

    /** 总容量文本 */
    val totalSizeText: String
        get() = Song.formatSize(totalSize)

    /** 已用容量文本 */
    val usedSizeText: String
        get() = Song.formatSize(usedSize)

    /** 可用容量文本 */
    val freeSizeText: String
        get() = Song.formatSize(freeSize)

    companion object {
        fun parse(json: String): StorageInfo? {
            if (json.isBlank() || json.trim() == "{}") return null
            val o = JSONObject(json)
            return StorageInfo(
                totalSize = o.optInt("totalSize"),
                usedSize = o.optInt("usedSize"),
                freeSize = o.optInt("freeSize"),
                systemSize = o.optInt("systemSize"),
                name = o.optString("name"),
                model = o.optString("model"),
                isPresent = o.optBoolean("isPresent")
            )
        }
    }
}
